crates/pallet-bioauth/src/migrations/consumed_auth_ticket_nonces_cleaner.rs

```diff
@@ -0,0 +1,70 @@
+//! Migration to clean consumed auth ticket nonces except ones related to current generation.
+
+#[cfg(feature = "try-runtime")]
+use frame_support::sp_std::vec::Vec;
+use frame_support::{log::info, pallet_prelude::*, traits::OnRuntimeUpgrade};
+
+use crate::{Config, ConsumedAuthTicketNonces, Pallet};
+
+/// Execute migration to clean consumed auth ticket nonces except ones related to current generation.
+pub struct ConsumedAuthTicketNoncesCleaner<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for ConsumedAuthTicketNoncesCleaner<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let pallet_name = Pallet::<T>::name();
+        info!("{pallet_name}: Running migration to clean consumed auth ticket nonces except ones related to current generation");
+
+        ConsumedAuthTicketNonces::<T>::mutate(|consumed_auth_ticket_nonces| {
+            if let Some(last_consumed_auth_ticket_nonce) = consumed_auth_ticket_nonces.last() {
+                let consumed_auth_ticket_nonces_number_before = consumed_auth_ticket_nonces.len();
+
+                // Current generation is represented by first 16 bytes at nonce.
+                let current_generation = last_consumed_auth_ticket_nonce[..16].to_vec();
+
+                consumed_auth_ticket_nonces.retain(|consumed_auth_ticket_nonce| {
+                    consumed_auth_ticket_nonce.starts_with(&current_generation)
+                });
+
+                let removed_number = consumed_auth_ticket_nonces_number_before
+                    .checked_sub(consumed_auth_ticket_nonces.len())
+                    .expect("valid operation; qed");
+
+                info!("{pallet_name}: Removed {removed_number} consumed auth ticket nonces");
+            } else {
+                info!("{pallet_name}: Nothing to remove");
+            }
+        });
+
+        T::DbWeight::get().reads_writes(1, 1)
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        let pre_upgrade_state = match ConsumedAuthTicketNonces::<T>::get().last() {
+            Some(last_consumed_auth_ticket_nonce) => {
+                last_consumed_auth_ticket_nonce.clone()[..16].to_vec()
+            }
+            None => Vec::new(),
```

Comment:
Perhaps it would be worth describing the type of this value more explicitly? E.g.

```rs
#[cfg(feature = "try-runtime")]
#[derive(Decode, Encode)]
struct PreUpgradeState {
    current_generation: Option<[u8; 16]>,
}
```

Reply:
Thanks for noting, forgot to add comment.

Anyway, I would prefer to use just comment as we usually did before if the data is quite simple.

Reply:
https://github.com/humanode-network/humanode/pull/1436/commits/1228f36c6f1157faf47dad6f24057016258b6d66